File: app/src/main/java/net/mbonnin/arcanetracker/ViewManager.java
package net.mbonnin.arcanetracker;

import android.app.Activity;
import android.content.Context;
import android.graphics.PixelFormat;
import android.graphics.Point;
import android.os.Build;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.view.WindowManager;

import java.util.ArrayList;
import java.util.Iterator;

/**
 * Created by martin on 10/21/16.
 */

public class ViewManager {
    public static ViewManager sViewManager;
    private final int mScreenWidth;
    private final int mScreenHeight;
    private final int mUsableWidth;
    private final int mUsableHeight;
    private WindowManager mWindowManager;
    private ArrayList<View> mViews = new ArrayList<>();

    public int getUsableWidth() {
        return mUsableWidth;
    }

    public int getUsableHeight() {
        return mUsableHeight;
    }

    public void addCenteredView(View view) {
        Params params = new Params();

        int wMeasureSpec = View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED);
        view.measure(wMeasureSpec, wMeasureSpec);

        params.x = (mScreenWidth - view.getMeasuredWidth()) / 2;
        params.y = (mScreenHeight - view.getMeasuredHeight()) / 2;
        params.w = view.getMeasuredWidth();
        params.h = view.getMeasuredHeight();

        addModalAndFocusableView(view, params);
    }

    public void removeAllViewsExcept(View view) {
        Iterator<View> it = mViews.iterator();
        while (it.hasNext()) {
            View view2 = it.next();
            if (view != view2) {
                it.remove();
                mWindowManager.removeView(view2);
            }
        }
    }

    public static class Params {
        public int x, y, w, h;
    }

    public ViewManager(Context context) {
        mWindowManager = (WindowManager)context.getSystemService(Activity.WINDOW_SERVICE);
        Point screenSize = new Point();

        mWindowManager.getDefaultDisplay().getSize(screenSize);
        if (screenSize.x > screenSize.y) {
            mUsableWidth = screenSize.x;
            mUsableHeight = screenSize.y;
        } else {
            mUsableHeight = screenSize.x;
            mUsableWidth = screenSize.y;
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            mWindowManager.getDefaultDisplay().getRealSize(screenSize);
            if (screenSize.x > screenSize.y) {
                mScreenWidth = screenSize.x;
                mScreenHeight = screenSize.y;
            } else {
                mScreenHeight = screenSize.x;
                mScreenWidth = screenSize.y;
            }
        } else {
            /**
             * best effort
             */
            mScreenWidth = mUsableWidth;
            mScreenHeight = mUsableHeight;
        }
    }

    public static  ViewManager get() {
        if (sViewManager == null) {
            sViewManager = new ViewManager(ArcaneTrackerApplication.getContext());
        }

        return sViewManager;
    }

    public void addModalAndFocusableView(View view, Params params) {
        addModalViewIternal(view, params, 0);
    }
    public void addModalViewIternal(View view, Params params, int extraFlags) {
        addViewInternal(view, params, WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL |  WindowManager.LayoutParams.FLAG_WATCH_OUTSIDE_TOUCH | extraFlags);
        view.setOnTouchListener((v, event) -> {
            if (event.getActionMasked() == MotionEvent.ACTION_OUTSIDE) {
                removeView(view);
            }
            return false;
        });
    }
    public void addModalView(View view, Params params) {
        addModalViewIternal(view, params, WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE);
    }
    private void addViewInternal(View view, Params params, int flags) {
        if (mViews.contains(view)) {
            return;
        }
        WindowManager.LayoutParams layoutParams = new WindowManager.LayoutParams(
                params.w,
                params.h,
                WindowManager.LayoutParams.TYPE_PHONE,
                flags,
                PixelFormat.TRANSLUCENT);
        layoutParams.gravity = Gravity.TOP | Gravity.LEFT;
        layoutParams.x = params.x;
        layoutParams.y = params.y;

        mViews.add(view);
        mWindowManager.addView(view, layoutParams);

    }
    public void addView(View view, Params params) {
        addViewInternal(view, params, WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE);
    }

    public void removeView(View view) {
        if (!mViews.contains(view)) {
            return;
        }
        mViews.remove(view);
        mWindowManager.removeView(view);
    }
    public void updateView(View view, Params r) {
        if (!mViews.contains(view)) {
            return;
        }

        WindowManager.LayoutParams layoutParams = (WindowManager.LayoutParams) view.getLayoutParams();
        layoutParams.width = r.w;
        layoutParams.height = r.h;
        layoutParams.x = r.x;
        layoutParams.y = r.y;

        mWindowManager.updateViewLayout(view, layoutParams);

    }

    public int getWidth() {
        return mScreenWidth;
    }

    public int getHeight() {
        return mScreenHeight;
    }

    public void removeAllViews() {
        while (!mViews.isEmpty()) {
            View view = mViews.get(0);
            removeView(view);
        }
    }

}

File: app/src/main/java/net/mbonnin/arcanetracker/ViewManager.kt
package net.mbonnin.arcanetracker

import android.app.Activity
import android.content.Context
import android.graphics.PixelFormat
import android.graphics.Point
import android.os.Build
import android.view.*
import com.crashlytics.android.Crashlytics
import java.util.*

/**
 * Created by martin on 10/21/16.
 */

class ViewManager(context: Context) {
    val width: Int
    val height: Int
    val usableWidth: Int
    val usableHeight: Int
    private val mWindowManager: WindowManager
    private val mViews = ArrayList<View>()

    fun addCenteredView(view: View, modal: Boolean = true) {
        val params = Params()

        val wMeasureSpec = View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED)
        view.measure(wMeasureSpec, wMeasureSpec)

        params.x = (width - view.measuredWidth) / 2
        params.y = (height - view.measuredHeight) / 2
        params.w = view.measuredWidth
        params.h = view.measuredHeight

        if (modal) {
            addModalAndFocusableView(view, params)
        } else {
            addView(view, params)
        }
    }

    fun removeAllViewsExcept(view: View) {
        val it = mViews.iterator()
        while (it.hasNext()) {
            val view2 = it.next()
            if (view !== view2) {
                it.remove()
                mWindowManager.removeView(view2)
            }
        }
    }

    class Params {
        var x: Int = 0
        var y: Int = 0
        var w: Int = 0
        var h: Int = 0
    }

    init {
        mWindowManager = context.getSystemService(Activity.WINDOW_SERVICE) as WindowManager
        val screenSize = Point()

        mWindowManager.defaultDisplay.getSize(screenSize)
        if (screenSize.x > screenSize.y) {
            usableWidth = screenSize.x
            usableHeight = screenSize.y
        } else {
            usableHeight = screenSize.x
            usableWidth = screenSize.y
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            mWindowManager.defaultDisplay.getRealSize(screenSize)
            if (screenSize.x > screenSize.y) {
                width = screenSize.x
                height = screenSize.y
            } else {
                height = screenSize.x
                width = screenSize.y
            }
        } else {
            /**
             * best effort
             */
            width = usableWidth
            height = usableHeight
        }
    }

    fun addModalAndFocusableView(view: View, params: Params) {
        addModalViewInternal(view, params, 0)
    }

    fun addModalViewInternal(view: View, params: Params, extraFlags: Int) {
        addViewInternal(view, params, WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL or WindowManager.LayoutParams.FLAG_WATCH_OUTSIDE_TOUCH or extraFlags)
        view.setOnTouchListener { _, event ->
            if (event.actionMasked == MotionEvent.ACTION_OUTSIDE) {
                removeView(view)
            }
            false
        }
    }

    fun addModalView(view: View, params: Params) {
        addModalViewInternal(view, params, WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE)
    }

    private fun addViewInternal(view: View, params: Params, flags: Int) {
        if (mViews.contains(view)) {
            return
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            Crashlytics.setBool("canDrawOverlay", android.provider.Settings.canDrawOverlays(view.context))
        }

        val layoutParams = WindowManager.LayoutParams(
                params.w,
                params.h,
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY else WindowManager.LayoutParams.TYPE_PHONE,
                flags,
                PixelFormat.TRANSLUCENT)
        layoutParams.gravity = Gravity.TOP or Gravity.LEFT
        layoutParams.x = params.x
        layoutParams.y = params.y

        mWindowManager.addView(view, layoutParams)
        mViews.add(view)
    }

    fun addView(view: View, params: Params) {
        addViewInternal(view, params, WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE)
    }

    fun removeView(view: View) {
        if (!mViews.contains(view)) {
            return
        }
        mViews.remove(view)
        mWindowManager.removeView(view)
    }

    fun updateView(view: View, r: Params) {
        if (!mViews.contains(view)) {
            return
        }

        val layoutParams = view.layoutParams as WindowManager.LayoutParams
        layoutParams.width = r.w
        layoutParams.height = r.h
        layoutParams.x = r.x
        layoutParams.y = r.y

        mWindowManager.updateViewLayout(view, layoutParams)
    }

    fun addViewWithAnchor(view: View, x: Int, y: Int) {
        val params = ViewManager.Params()

        val measureSpec = View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED)
        view.measure(measureSpec, measureSpec)
        params.w = view.getMeasuredWidth()
        params.h = view.getMeasuredHeight()

        params.x = (x + Utils.dpToPx(20))
        params.y = (y - params.h / 2)
        if (params.y < 0) {
            params.y = 0
        } else if (params.y + params.h > ViewManager.Companion.get().height) {
            params.y = ViewManager.Companion.get().height - params.h
        }

        addView(view, params)
    }

    fun removeAllViews() {
        while (!mViews.isEmpty()) {
            val view = mViews[0]
            removeView(view)
        }
    }

    fun allViews(): List<View> {
        return mViews
    }

    fun addMenu(menuView: View, anchorView: View) {
        menuView.layoutParams = ViewGroup.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        val wMeasureSpec = View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED)
        menuView.measure(wMeasureSpec, wMeasureSpec)

        val a = IntArray(2)
        anchorView.getLocationOnScreen(a)

        val params = ViewManager.Params()
        params.x = a[0] + anchorView.width / 2
        params.y = a[1] + anchorView.height / 2 - menuView.measuredHeight
        params.w = menuView.measuredWidth
        params.h = menuView.measuredHeight

        addModalView(menuView, params)

    }

    companion object {
        var sViewManager: ViewManager? = null

        fun get(): ViewManager {
            if (sViewManager == null) {
                sViewManager = ViewManager(ArcaneTrackerApplication.context)
            }

            return sViewManager!!
        }
    }

}
